### ISSUE TITLE
Collapse all subclasses of Partvalue into just Partvalue

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#329]
The sublcasses DockingPortValue, EngineValue, and SensorValue, 
of PartValue

Can be just collapsed down into PartValue and use fields to decide which is which,like for example:

PART:ISENGINE()

PART:ISDOCKINGPORT()

etc.

This is closer to how the actual KSP API does it.

**Comment by erendrake** [comment]
you are right, this not a job for inheritance. Squad moved away from having them as inherited types for just this same reason

**Comment by AlchemistCH** [comment]
I'll leave here some description of my current concept
https://github.com/AlchemistCH/KOS/blob/PartOverhaul/

For end-user it's mostly like described in this issue, but source-code wise it's a multilevel structure:

Basic part functionality is handled by PartValue class, but it's used as an abstract.
The class that the user will get as Part is PartValueExt, derived from PartValue.
PartValueExt (extendable) class handles attachment of module-based functionality to the part by creating the extenders for those classes and making them to put suffixes back on itself.
This suffix dumping in case of single functional module allows to access all the functionality from the part itself. In more complex cases the extender logic may be a bit different.

PartExtender (or, to be precise, it's derivatives designed each for specific class functionality) is middle level class handling a group of modules. The typical case (a group of independently functional part modules) is handled like this (the example contains such system for resource converters):
All suffixes in part-based case are actually to be paced on the part, so that the end-user won't have to reference the extender.
Separate modules are accessible as a list of ModuleFields-derived objects designed for specific module. Also a search function may be present.
In the typical cases the extender copies the suffixes of the corresponding ModuleFields subclass, but those are made to handle all the corresponding modules. 
For single-module case the same suffixes can be used from the Part and from the partModule with the same result.
On the other hand, the handle-all-modules functionality may be useful on ship scale (stop all converters, run all experiments, extend all solar panels). Therefore the extenders have an alternate constructor working with a list of parts (in this case the extender puts the suffixes on itself).

An exception to this concept is the case when simultaneous handling of a bunch of modules is not logical (which turned out to be both most used already existing classes):
Docking ports definitely have to be addressed on a single-port basis, so the DockingPortExt in case of multiple modules allows choosing the one referenced by the part-level suffixes. (On the other hand, I don't think stock even can properly support multiple docking ports on a single part at this point)
Engine (with regard to how the multi-mode case works, uniting 2 separate engine modules under control of a third module into single user-accessible engine) is something that's better to handle on part basis, not module basis, therefore it's handled by the good old (but slightly modified) EngineValue object. It copies all the suffixes on the Part level, but (in case of suffix mess by multitude of modules) is also fully accessible through Part:ENGINE.

Additional note: some of the interaction between different modules is to be handled on module class basis. As example, DockingPortFields got a few suffixes that control the attached animation module (if present) for opening animated ports. On the other hand, converter modules already have their interaction with core heat module covered by their own functions.
